### ISSUE TITLE
Off-by-one-CC-num sent in `Utility/TimeWaster MIDI-Mouse.jsfx`

### ISSUE THREAD

**Issue by GezzMC** [ReaTeam/JSFX#121]
Thanks for collecting other people's JSFX into one repository. 

I'm thinking that you'd prefer this be reported on https://forum.cockos.com/showthread.php?t=169127 (link in your README), but I'm here as a user, have a bug report not a question, and don't have a REAPER forum account. I say more on this below.

Script `TimeWaster MIDI-Mouse.jsfx` is very useful, but it sends a CC that's one greater than is specified by the slider.

The problem is line 35:

    ccmsb = (cctarget-1) <= 31 ? cctarget:cctarget+32;

The assignment should result in `ccmsb = (cctarget - 1)`.

That's the only problem I know of, and now I express some opinions. To make it clear, there are no problems. If you change the script in a way that makes it good for me, I'll use the `ReaTeam/JSFX` version. If you don't, I'll use the script as I've modified it, which I may distribute with my projects.

I use the JSFX script for basic testing, and Malcolm Smith, the scripter, is sending CC in a non-conventional way. 

* Per the MIDI standard, CC#32-63 are tagged as LSB for CC#0-31, but that's not how the CC are conventionally used.
* For example, CC1, the modwheel, doesn't get sent out with CC#33 by keyboards.
* So CC LSB getting sent out ends up as clutter when I'm looking at a MIDI monitor, and if I were to assign something else to CC#33, such as a synth parameter, it would make the JSFX script unusable.

Well, I modified things to suit my purposes. If you want to see what I did, I can post my modification here, but GitHub doesn't seem to partially show a long piece of code, with a scroll bar.

### Bug reporting etiquette ###

Maybe use of REAPER scripting is growing and will explode in use, and there will be some growing pains on how things are done.

My general rule is to stay out of forums and limit the number of web accounts that I use. And `TimeWaster MIDI-Mouse.jsfx` is distributed on GitHub, and GitHub has the `Issues` tab.

I do notice I'm the first to report a problem here. If you only want people to report stuff on the forum, you might consider making that know explicitly in the README, and maybe even disabling `Issues`.

Thanks. I don't know that you can delete this, but if you can and want to, go ahead.

**Comment by cfillion** [comment]
Pinging @the-wasters-of-time, the author of that effect. Also that file is marked as noindex without being included in any ReaPack packages, and so it should probably be removed from this repository.

**Comment by GezzMC** [comment]
Thanks for the reply.

This reply of mine will mainly be my opinions about the world of JSFX scripts, how things work, and how things would work in a more perfect world. I'm not a pro in any way, but most of what I say are pretty basic programming principles.

If you don't close this issue, I'll close it in a day or two.

My presense here is most likely a one-shot deal, unless you give directions in your README that issues should be reported here, not that I assume I'll find bugs in the select few scripts I use from here.

Here are the topics below: 

* Not replying direct to the script writer, and the `MIDI-Mouse` bug was fixed elsewhere.
* User TiaR (AKA SmashedTransistors) has lots of JSFX in `ReaTeam/JSFX`, but his amazing  `TwoStageDist_02.jsfx` isn't in the set.
  * This is an example of the haphazardness of JSFX distribution (the fun is the initial creation, after that it's tedious work).
* The old-skool haphazardness of JSFX script versioning, and script namespace as "filename of the script in the `REAPER/Effects` folder" (what I've discovered based on my tests).
* As insiders or people with sway, please get a feature added to REAPER where there's a local `Effects` folder for a RPP project, that gets scanned before `REAPER/Effects`.
  * Projects could then be self-contained as far as JSFX, and this could guarantee that there aren't any JSFX filename conflicts.

### Not replying direct, from my rules for the web ###

As to replying to @the-wasters-of-time, I prefer to maintain somewhat proper protocol. If someone has a repository and has `Issues` open, I might notify them, because if I'm using their script, it must be of value. But otherwise, I just try to meet my own needs. 

Malcolm doesn't have a repository that contains the particular JSFX script, so I don't reply to him. He fixed the problem already anyway, but on [stash.reaper.fm](https://stash.reaper.fm).

I had to search for that because he didn't put a reference URL in the JSFX script, unlike most people. Not a problem. Script writers are free to do what they want.

### Classic result of having multiple locations for the same file ###

In `MIDI-Mouse`, there's no referral link. I've now done the searches to find things:

* https://stash.reaper.fm/v/33047/TimeWaster%20MIDI-Mouse.jsfx
* https://forum.cockos.com/showthread.php?t=204105

Malcom fixed the problem in his [stash.reaper.fm version](https://stash.reaper.fm/v/33047/TimeWaster%20MIDI-Mouse.jsfx), but his `ReaTeam/JSFX` copy didn't get updated. 

Doesn't this represent part of the madness of JSFX scripts? Links to scripts can be in multiple places, with it not being obvious what's the latest version.
  
Much of the time there are multiple links on a [forum.cockos.com](https://forum.cockos.com) thread, leading to various ways the script is hosted:

* on GitHub,
* on [stash.reaper.fm](https://stash.reaper.fm),
* on a person's own web site,
* where these different locations may or may not host a mirror of the script.

I've probably not said anything that you haven't already thought about, since there is, after all, `ReaTeam/JSFX`.

### `ReaTeam/JSFX`: example of "more proper", simplicity for redistribution ###

I've been watching REAPER for a while, and this https://github.com/ReaTeam/JSFX seems like a big step in the right direction of trying to bring method to the madness of how JSFX scripts are distributed and maintained. 

There is also https://reapack.com, but I don't use that because I prefer to micromanage my dependencies, and none of this has anything to do with my own needs. I only care about all this because, in the future, I want to distribute a small number of JSFX scripts with my RPP projects, and give people a one-step instruction on how to get the scripts to the `REAPER/Effects` directory.

This brings me back to `ReaTeam/JSFX`. When looking for scripts, I first look in your set of scripts, to try to reduce the number of site links I will need to provide, so people can go to the original source. And a download from GitHub is better than a person searching through a forum thread, looking for the right link to click on, to get the newest version of a JSFX script.

### User T.Rochebois/TiaR/SmashedTransisters and `TwoStageDist_02.jsfx` ###

User `TiaR`, as he's called in `ReaTeam/JSFX`, and his scripts are an example of many things in relation to JSFX scripts, part of that being `ReaTeam/JSFX`.

The thread where I got the 2-stage distortion script is https://forum.cockos.com/showthread.php?t=183376.

My search for a JSFX distortion script has been very narrow and specific. A good VST is [Disto::FX](http://distocore.net/disto_fx.html), but I've gotten rid of all VSTs except for [Dexed](https://asb2m10.github.io/dexed/). I did that by finally learning how to use the non-graphical displays of JSFX plugins, and sorting through JSFX scripts, to find what was is both good and approximates the good things I was getting out of particular VST plugins.

It's probably not obvious, but the main point in this section is that JSFX scripts are huge in REAPER being a lightweight DAW setup. Little JSFX text scripts do amazing things, they help keep the setup simple, and if distributed with a RPP file, they add nothing much to the size of a ZIP file.

Anyway, SmashedTransister left `TwoStageDist_02.jsfx` at `beta 02`, but it's so far working good for me, though I've only done initial tests. I've sorted through lots of distortion plugins, VST and JSFX, and so far it's a keeper for me, so it seems it should be in `ReaTeam/JSFX`.

It has a nice graphical display that shows the distortion curves, and the output waveforms.

### JSFX namespacing and versioning ###

I guess JSFX kind of ended up like JavaScript. It started out as a simple solution, then got popular, and is now used in bigger ways than maybe planned.

It's not like developers need me to say this, but lots of people writing 3rd-party JSFX scripts guarantees namespace conflicts.

It's my next request that would allow a person to guarantee that no JSFX filename conflicts occur due to filename conflicts.

### Local project `Effects` folder scanned ###

No DAW is perfect, but REAPER is a lightweight download and install, and the use of JSFX scripts can be exploited to keep things lightweight.

My plan is to distribute the few select JSFX scripts that I use, 6-10 of them, with my REAPER project. I'll have a README that says, "Extract the JSFX scripts to your `REAPER/Effects` and open the project. It should work if you know the basic of how to use REAPER."

Maybe there's also Dexed, but if there's only JSFX scripts, then things are simple.

Additionally, I guarantee that there's little chance of filename conflicts, by prefixing other people's (3rd-party) scripts with something like `190501gezzmc__foo.jsfx`. Ugly, and highly undesirable, but the best solution I can think of.

If there was a project `Effects` folder that got scanned, I wouldn't even have to tell people to unzip the scripts to `REAPER/Effects`.

So, as insiders, I'm depending on you to get this feature for me. That's a joke, sort of.

### Issue comment termination ###

Thanks for your time. Like I said, I'll close this issue if you don't. Maybe tomorrow.

**Comment by cfillion** [comment]
ReaTeam/JSFX is the stock effect repository for ReaPack. Effect developers upload and update their own work here to make it available for installation within the extension (in addition to making it available for manual download on GitHub–though that's not recommended as some packages here consist of many files, updating must be done by hand...).

If you want something to be added to this repository or updated, point the developer to https://reapack.com/upload/jsfx.

> It's my next request that would allow a person to guarantee that no JSFX filename conflicts occur due to filename conflicts.

Not a problem here? Installing an effect package using ReaPack puts it in a per-repository subdirectory. And all packages within a single repository cannot conflict with another.

> If there was a project Effects folder that got scanned, I wouldn't even have to tell people to unzip the scripts to REAPER/Effects.

REAPER supports loading JSFX from a per-project directory since version 5.12.

**Comment by GezzMC** [comment]
### 1st thing 1st ###

At this point, there are no problems. Everything is good and has worked out better than I expected.

There was a bug in the old version here of `TimeWaster MIDI-Mouse.jsfx`, but if there wouldn't have been, I wouldn't have made modifications to it, and my mods suit my purposes better.

I've now done the experiments to see how a REAPER project loads a JSFX script from the project `Effects` folder. There are things to say about that, but I say them below. Knowing it can be done is a huge solution. Thank you very much for the info. On the forum maybe I would have gotten that info sooner, but maybe not. In the past and even today, I've done many searches in the manual and on the web about where to put JSFX scripts.

If you stop reading now, nothing of great value will be lost. But I write more.

### Context: magic to deal with complexity vs. eliminating complexity ###

To understand what I'm trying to achieve, you have to understand the context I'm working under.

There are two ways to deal with complexity. Creating magic tools is one way. Eliminating complexity is the other way. You're helping with both, but your focus is on the magic tool you've created or helped create, https://reapack.com.

I'm not just trying to make good music and distribute MP3 files on various sites. Amazing tools are now a dime a dozen, and the problem of distribution has long been solved.

I'm trying to make good music, give away my projects, and increase the chances that other people can easily duplicate my REAPER setup, so they can easily open my RPP file, hit the space bar, and have things play. (If a person can't get past "Hello, World!" quickly, there's a good chance they'll move on to something else quickly.)

Old-skool making music, which still dominates, is to grab whatever DAWs, VSTs, loop libraries, and sample libraries you want, commercial or free, legitimately obtained or not, and make music, good or not.

So, a musician typically has some setup that no other person on the face of the Earth has, and if the musician didn't take notes on how they set things up, even they would probably have a problem setting things up again.

I want to reduce complexity for both myself, and so that other people can reasonably duplicate my setup. Every single thing that I tell people to install will greatly reduce their willingness to install something for me.

This applies even to tech-skilled musicians, not just to newbies. A person who uses Ableton Live to make great music, but not REAPER, will not be willing to put much effort into setting up REAPER, even if they hear something I've done that interests them, even if I give them my project. On the other hand, if my setup is simple, and "it just works", they might be willing to load my project.

### The rest would be a book ###

To explain more would be 3000 words or so, and this is a bug report. I try to get brief now.

### Filename conflicts ###

You say: 

> Not a problem here? 

It's the wild west in how JSFX scripts are distributed, and a whole lot of people have written and are going to write worthy JSFX scripts that aren't distributed with ReaPack. I already mentioned one, `TwoStageDist_02.jsfx`.

But people's problems that'll arise due to JSFX filename conflicts aren't my problem. It just ends up being one of numerous things people need to learn about, on the huge learning curve of how to use REAPER.

### Submitting scripts for ReaPack ###

Unfortunately, I'm having to come up to speed on REAPER script writing, to write a JSFX to trigger pitch bend curves with CC, to hopefully make it easy to emulate guitar-style vibrato, and much much more.

In a more perfect world, I would put out the call on https://forum.cockos.com asking for skilled-JSFX-developers to do this, but I don't operate on forums anymore.

I guess this ties back into that some people are going to write scripts, but not submit them for ReaPack. It's an imperfect world.

### Most importantly, per-project scripts ###

Looking through the manual, I found the appropriate place this should be noted, section 6.25, *Installing Extra JS Plug-ins*. Maybe someone on the inside, or with influence, can edit the manual to let others know. But I don't care, now that I know.

I've learned some details of how REAPER searchs for JSFX scripts by experimenting, moving scripts around, and generally trying to figure out what breaks things and what doesn't. Here's a line from my `.rpp` project template:

```
<JS "!gezzmc/DIR NO MATTER/gezzM1V1__TwoStageDist_02.jsfx"
```

In the process of editing my RPP to fix the mysterious problem of FX plugin windows getting stuck at a fixed window dimension, I noticed many old paths to JSFX. Rather than correct them, I changed the old paths to `DIR NO MATTER`. Maybe it'll break things in the future. It's not good to put trust in information that doesn't come from authoritative sources, at least with software.

And just from today, here's a line in my RPP for testing how project `Effects` works:


```
<JS <Project>/TEST_DIR/TEST__TwoStageDist_02.jsfx
```

But the script isn't in the `TEST_DIR` folder now, though the RPP still loads fine. It finds any script that's in the `Effects` folder of the root folder of the project. Another example of magic. There's a lot of magic with a default REAPER install.

However, a script can't simply be moved from `REAPER/Effects` to `fooproject/Effects`. 

In my case, there are no problems, there are only workarounds. When I get ready to distribute a REAPER project, I'll just copy the JSFX scripts to the project `Effects`, and edit my RPP file to change the paths to `<Project>`.

I'm thinking this would be useful information for other people to find through searches.

### Magic ReaPack vs. no magic ###

There's no conflict here. Different people like to do things different. Options are good many times. Much of the time, it's good when it's not either/or, Mac vs. PC.

Thanks for the info.

**Comment by GezzMC** [comment]
I finally noticed that you're the only contributor to ReaPack, but my motivation is low to install it because I've already done much of the work to find JSFX to supplement what the REAPER distribution doesn't give me. 

I think it's good when people are attacking the problem from both ends. 

I've learned enough on the audio side to not really want a massive number of plugins, with massive bells and whistles. Delay, reverb, chorus, transparent brick wall limiter, eq, o'scope, and a few more audio and MIDI utilities, I've either learned that the REAPER stuff is actually very good, or accepted the limitations for the utility stuff.

I'm sure ReaPack would have helped when I knew less, and before I had gone through all the web searches. There are things I think about searching for, but however you do it, it takes time and work to sort through things to find the good stuff.

Thanks again.
